infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesNamespaceFactory.java

```diff
@@ -72,32 +88,92 @@ public KubernetesNamespaceFactory(
+  }
+
+  /** True if namespace is potentially created for the workspaces, false otherwise. */
+  protected boolean isCreatingNamespaces(String workspaceId) throws InfrastructureException {
+    // ...namespace    | ...namespace exists | ...namespace.default | creating?
+    // no-placeholders |       no            |       null           | error
+    // no-placeholders |       no            |   no-placeholders    | no
+    // no-placeholders |       no            |    placeholders      | yes
+    // no-placeholders |      yes            |       null           | no
+    // no-placeholders |      yes            |   no-placeholders    | no
+    // no-placeholders |      yes            |    placeholders      | no
+    //  placeholders   |       no            |        null          | error
+    //  placeholders   |       no            |   no-placeholders    | no
+    //  placeholders   |       no            |    placeholders      | yes
+    //  placeholders   |      yes            |        null          | yes
+    //  placeholders   |      yes            |   no-placeholders    | yes
+    //  placeholders   |      yes            |    placeholders      | yes
+    boolean legacyWithPlaceholders = isNullOrEmpty(namespaceName) || hasPlaceholders(namespaceName);
+    boolean legacyExists =
+        checkNamespaceExists(
+            resolveLegacyNamespaceName(EnvironmentContext.getCurrent().getSubject(), workspaceId));
+
+    if (isNullOrEmpty(defaultNamespaceName) && !legacyExists) {
+      throw new InfrastructureException(
+          "Cannot determine whether a new namespace and service account should be"
+              + " created for workspace %s. There is no pre-existing workspace namespace to be found using the legacy"
+              + " `che.infra.kubernetes.namespace` property yet the `che.infra.kubernetes.namespace.default` property"
+              + " is undefined.");
+    }
+
+    boolean defaultHasPlaceholders = hasPlaceholders(defaultNamespaceName);
+
+    return (legacyWithPlaceholders && legacyExists) || (!legacyExists && defaultHasPlaceholders);
   }
 
   /**
-   * True if namespace is predefined for all workspaces. False if each workspace will be provided
-   * with a new namespace or provided for each user when using placeholders.
+   * @return true if the namespaces are fully managed by Che (e.g. created, used and deleted), false
+   *     otherwise
    */
-  public boolean isPredefined() {
-    return isPredefined;
+  public boolean isManagingNamespaces(String workspaceId) throws InfrastructureException {
+    boolean canBeManaged =
+        isNullOrEmpty(namespaceName)
+            ? defaultNamespaceName != null && defaultNamespaceName.contains(WORKSPACEID_PLACEHOLDER)
+            : namespaceName.contains(WORKSPACEID_PLACEHOLDER);
+
+    if (!canBeManaged) {
+      return false;
+    }
+
+    if (isNullOrEmpty(namespaceName)) {
+      // if we're using the new logic, we don't have to check if the namespace exists
+      return true;
```

Comment:
null or empty actually means that `<workspace_id>` is used, I don't really understand why we should not check namespace existence here.

Reply:
Because it is superfluous. We only use the legacy logic for namespaces that are already in existence. We use the new logic in all other cases.